discovery-provider/tests/test_index_user_listening_history.py

```diff
@@ -0,0 +1,327 @@
+
+            {"item_id": 1, "user_id":3, "created_at": TIMESTAMP_3},
+
+            # New play
+            {"item_id": 2, "user_id":1, "created_at": TIMESTAMP_3}, # listen to new track
+            {"item_id": 1, "user_id":1, "created_at": TIMESTAMP_4}, # re-listen to existing track, dedupe
+
+        ] + [
+            # new user listens to many tracks
+            {"item_id": i+1, "user_id":4, "created_at": TIMESTAMP_4 + timedelta(hours=i)} for i in range(200)
+        ],
+
+        "indexing_checkpoints": [
+            {"tablename": USER_LISTENING_HISTORY_TABLE_NAME, "last_checkpoint": 6}
+        ]
+    }
+
+    populate_mock_db(db, entities)
+
+    with db.scoped_session() as session:
+        _index_user_listening_history(session)
+
+    with db.scoped_session() as session:
+        results: List[UserListeningHistory] = (
+            session.query(UserListeningHistory)
+            .order_by(UserListeningHistory.user_id)
+            .all()
+        )
+
+        assert len(results) == 4
+
+        assert results[0].user_id == 1
+        assert len(results[0].listening_history) == 2
+        assert results[0].listening_history[0]["track_id"] == 1
+        assert results[0].listening_history[0]["timestamp"] == str(TIMESTAMP_4)
+        assert results[0].listening_history[1]["track_id"] == 2
+        assert results[0].listening_history[1]["timestamp"] == str(TIMESTAMP_3)
+
+        assert results[1].user_id == 2
+        assert len(results[1].listening_history) == 2
+        assert results[1].listening_history[0]["track_id"] == 2
+        assert results[1].listening_history[0]["timestamp"] == str(TIMESTAMP_2)
+        assert results[1].listening_history[1]["track_id"] == 1
+        assert results[1].listening_history[1]["timestamp"] == str(TIMESTAMP_1)
+
+        assert results[2].user_id == 3
+        assert len(results[2].listening_history) == 3
+        assert results[2].listening_history[0]["track_id"] == 1
+        assert results[2].listening_history[0]["timestamp"] == str(TIMESTAMP_3)
+        assert results[2].listening_history[1]["track_id"] == 3
+        assert results[2].listening_history[1]["timestamp"] == str(TIMESTAMP_2)
+        assert results[2].listening_history[2]["track_id"] == 2
+        assert results[2].listening_history[2]["timestamp"] == str(TIMESTAMP_1)
+
+        assert results[3].user_id == 4
+        assert len(results[3].listening_history) == 100
+        for i in range(100):
+            assert results[3].listening_history[i]["track_id"] == 200 - i
+            assert (
+                results[3].listening_history[i]["timestamp"] ==
+                str(datetime.fromisoformat("2014-04-12 07:00:00") - timedelta(hours=i))
+            )
+
+        new_checkpoint: IndexingCheckpoints = (
+            session.query(IndexingCheckpoints.last_checkpoint)
+            .filter(IndexingCheckpoints.tablename == USER_LISTENING_HISTORY_TABLE_NAME)
+            .scalar()
+        )
+
+        assert new_checkpoint == 208
```

Comment:
this test case just made me think about the Checkpoint table having this value as an Integer, which has a max of `2,147,483,647`

This is a lot, but it's not insane if we're hitting > 200K tracks/day. We may hit issues with this eventually. Thought i'd just call out (can ignore now). This would be ~5 years at 1MM plays/day.

Reply:
Ooh yeah... the id column in plays is limited by that as well. Maybe we should just address all of that later on.